### PR TITLE
Fix Promise::when invocation order

### PR DESCRIPTION
**Pull request by kelunik**

### CODE REVIEW

**Comment by trowski** [comment]
Unfortunately this breaks the protection of double resolving a promise. Now a promise can be resolved again in the when callback.

File: test/WhenOrderTest.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+class WhenOrderTest extends \PHPUnit_Framework_TestCase {
+    public function testWhenOrder() {
+        $this->expectOutputString("1234");
+
+        $deferred = new Deferred;
+        $promise = $deferred->promise();
+
+        $promise->when(function () use ($promise) {
+            $promise->when(function () {
```
Comment (joshdifabio) [inline test/WhenOrderTest.php]: This `when()` isn't invoked a synchronously with the current implementation. Is that your intention?

**Comment by joshdifabio** [comment]
How about updating the test to check which `when()` calls are handled synchronously?

**Comment by joshdifabio** [comment]
This is what I had in mind: https://github.com/amphp/amp/pull/58

File: test/WhenOrderTest.php (hunk 1 above)
Comment (kelunik) [inline test/WhenOrderTest.php]: I had not considered that. We have to define when a promise is resolved. I guess it should be _before_ the callbacks are invoked.

Comment (joshdifabio) [inline test/WhenOrderTest.php]: I agree. I think it has to be before, otherwise we have fired the listeners for a promise when it's still in _pending_ state.

**Comment by kelunik** [comment]
I think the current order is actually the right one and we should change the spec to account for that.

File: lib/Internal/Placeholder.php [hunk 2]
```diff
@@ -3,7 +3,8 @@
 namespace Amp\Internal;
 
 use Amp\Failure;
-use Interop\Async\{ Loop, Promise };
+use Interop\Async\Loop;
```
Comment (trowski) [inline lib/Internal/Placeholder.php]: IDE change this?

Comment (kelunik) [inline lib/Internal/Placeholder.php]: Both, IDE changed it to

```
use Interop\Async\{
    Loop, Promise
};
```

I changed it then to avoid those stupid group uses.

Comment (trowski) [inline lib/Internal/Placeholder.php]: That's annoying… probably isn't worth grouping them then unless there's 3 or more.

Comment (kelunik) [inline lib/Internal/Placeholder.php]: It's never worth. Group use shouldn't be a thing. It's way harder to read and find the right entries.

**Comment by joshdifabio** [comment]
Wait. Is it just me or is it impossible for the callbacks to be executed in the order they're added to the promise _and_ for `when()` calls on resolved promises to be fired synchronously? Looking at your test this seems to be the case.

**Comment by kelunik** [comment]
@joshdifabio Right.

**Comment by bwoebi** [comment]
@joshdifabio right. But that's an edge case though, I do not care too much; I think immediate resolution should take priority though (least surprise IMHO).

**Comment by bwoebi** [comment]
Can we please close that PR and defer this to async-interop/promise-test?

**Comment by kelunik** [comment]
@bwoebi No we can't. `async-interop/promise-test` only tests already resolved promises.

**Comment by bwoebi** [comment]
@kelunik How do you think that? … no?!?

**Comment by kelunik** [comment]
Nevermind.
